# Day 10: Model Deployment

### 🔹 Introduction
Today, we focus on **model deployment basics**.  

We will:  
1. Save our **best tuned Gradient Boosting model**.  
2. Load it back to confirm it works.  
3. Test predictions to verify the saved model behaves correctly.  



In [1]:
import joblib
from sklearn.metrics import mean_squared_error, r2_score

# Load train-test splits (from earlier saved files)
import joblib
X_train = joblib.load("X_train.pkl")
X_test = joblib.load("X_test.pkl")
y_train = joblib.load("y_train.pkl")
y_test = joblib.load("y_test.pkl")

# Best Gradient Boosting model from Day 9
from sklearn.ensemble import GradientBoostingRegressor
best_gb_model = GradientBoostingRegressor(
    learning_rate=0.05, max_depth=2, n_estimators=200, random_state=42
)
best_gb_model.fit(X_train, y_train)

# Save model
joblib.dump(best_gb_model, "best_gb_model.pkl")
print("✅ Best Gradient Boosting Model saved as best_gb_model.pkl")

# Load model back
loaded_model = joblib.load("best_gb_model.pkl")

# Evaluate loaded model
loaded_preds = loaded_model.predict(X_test)
mse = mean_squared_error(y_test, loaded_preds)
r2 = r2_score(y_test, loaded_preds)

print("✅ Loaded Model Test Performance:")
print("MSE:", round(mse, 2))
print("R²:", round(r2, 2))


✅ Best Gradient Boosting Model saved as best_gb_model.pkl
✅ Loaded Model Test Performance:
MSE: 0.21
R²: 0.98


### 🔹 Conclusion  

- We successfully **saved the best tuned Gradient Boosting model** as `best_gb_model.pkl`.  
- Reloading the model gave consistent results:  
  - **MSE: 0.21**  
  - **R²: 0.98**  
- This confirms our model can be stored and reused for future predictions without retraining.  
- Next steps could include integrating this model into a simple API or web app for real-world deployment. 🚀  
